homeassistant/components/esphomelib/__init__.py

```diff
@@ -0,0 +1,338 @@
+                         host, error)
+            # Schedule re-connect in event loop in order not to delay HA
+            # startup. First connect is scheduled in tracked tasks.
+            data['reconnect'] = hass.loop.create_task(
+                try_connect(tries + 1, is_disconnect))
+        else:
+            _LOGGER.info("Successfully connected to %s", host)
+
+    cli.on_disconnect = try_connect
+    return try_connect
+
+
+async def _cleanup_instance(hass: HomeAssistantType, entry: ConfigEntry):
+    """Cleanup the esphomelib client if it exists."""
+    data = hass.data[DOMAIN].pop(entry.entry_id)
+    client = data[DATA_CLIENT]
+    if DATA_RECONNECT_TASK in data:
+        data[DATA_RECONNECT_TASK].cancel()
+    await client.stop()
+
+
+async def async_unload_entry(hass, config_entry) -> bool:
+    """Unload an esphomelib config entry."""
+    await _cleanup_instance(hass, config_entry)
+
+    for component in HA_COMPONENTS:
+        await hass.config_entries.async_forward_entry_unload(
+            config_entry, component)
+
+    return True
+
+
+async def async_setup(hass, config):
+    """Stub to allow setting up this component.
+
+    Configuration through YAML is not supported at this time.
+    """
+    return True
+
+
+def get_entry_data(hass: HomeAssistantType,
+                   key: Union[ConfigEntry, str]) -> Dict[str, Any]:
+    """Get the per-config entry data dictionary for a key."""
+    if isinstance(key, ConfigEntry):
+        key = key.entry_id
+    return hass.data[DOMAIN].setdefault(key, {})
+
+
+async def platform_async_setup_entry(hass: HomeAssistantType,
+                                     entry: ConfigEntry,
+                                     async_add_entities,
+                                     *,
+                                     component_key,
+                                     info_type,
+                                     entity_type,
+                                     state_type,
+                                     ):
+    """Set up an esphomelib platform."""
+    get_entry_data(hass, entry).setdefault(component_key, {})
+
+    @callback
+    def async_list_entities(infos: List[Any]):
+        """Update entities of this platform when entities are listed."""
+        old_entities = get_entry_data(hass, entry)[component_key]
+        new_entities = {}
+        add_entities = []
+        for info in infos:
+            if not isinstance(info, info_type):
+                # Filter out infos that don't belong to this platform.
+                continue
+
+            if info.key in old_entities:
+                # Update existing entity
+                entity = old_entities.pop(info.key)
+            else:
+                # Create new entity
+                entity = entity_type(entry.entry_id)
+                add_entities.append(entity)
+            entity.async_info_update(info)
+            new_entities[info.key] = entity
```

Comment:
Don't store the entity, but store a reference to the entity, eg some unique id of the entity. We shouldn't store entities in shared containers. Entities are a platform concern.